overlapfinder_melvin.py

```diff
@@ -0,0 +1,130 @@
+from datetime import datetime as dt
+from intervaltree import Interval, IntervalTree
+#import test_inputs
+
+# one interval
+tc_0 = [Interval(1, 3, "aaron")]
+
+# no overlap
+tc_1 = [Interval(1, 2, "aaron"), Interval(3, 4, "ben"), Interval(5, 6, "charlotte")]
+
+# separate overlaps
+tc_2 = [Interval(1, 3, "aaron"), Interval(2, 4, "ben"),  Interval(3, 5, "charlotte")]
+
+# shared overlap
+tc_3 = [Interval(1, 3, "aaron"), Interval(2, 4, "ben"), Interval(2, 5, "charlotte")]
+
+# 4 intervals
+tc_4 = [Interval(1, 3, "aaron"), Interval(2, 4, "ben"),
+        Interval(3, 5, "charlotte"), Interval(1, 6, "dan")]
+
+
+
+# Input: An list of Intervals
+# Output: A dictionary (key: overlap, value: set of user_ids which share the overlap)
+def find_all_common_intervals(interval_list):
+    overlap_dict = dict()
+    interval_tree = IntervalTree(interval_list)
+    
+    for interval in interval_tree.items(): # create a copy of the tree to iterate over
+        # check interval against overlap_dict
+        for overlap in overlap_dict.copy():
+            add_overlap_to_dict(overlap, interval, overlap_dict)
+            
+        # check interval against other intervals in the tree to find overlapping regions
+        other_intervals = find_other_intervals_which_overlap(interval_tree, interval)
+        for other_interval in other_intervals:
+            add_new_overlap_to_dict(interval, other_interval, overlap_dict)
+               
+         
+    return overlap_dict
+
+
+# Input: An IntervalTree and an Interval in the IntervalTree
+# Output: A set containing all other Intervals in the IntervalTree that intersect with the given
+#  Interval
+def find_other_intervals_which_overlap(tree, interval):
+    tree.remove(interval)
+    result = tree.search(interval.begin, interval.end) # returns a set
+    return result
+
+
+# Input: Two intervals and a dictionary (key is Interval)
+# Output: None
+# Pre-condition: interval.data is a user id
+def add_new_overlap_to_dict(interval_a, interval_b, overlap_dict):
+    overlap = find_overlap(interval_a.begin, interval_a.end, interval_b.begin, interval_b.end) # returns a tuple
+    if overlap != None:
```

Comment:
I think the return is necessary.

As for the construction of the Interval object, I did it because previously the dictionary was keyed by tuple which is alright, but for consistency's sake, i changed the overlap_dict dictionary to be keyed by Interval objects intead.

Reply:
Oh, I see what you mean now, yes this is good. Forgot dictionary was keyed by tuples previously.